File: 백준/Platinum/1725. 히스토그램/히스토그램.kt
import java.util.*
import kotlin.math.max

fun main() {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt()
    val arr = LongArray(n + 2) { 0L }

    for (i in 1..n) {
        arr[i] = br.readLine().toLong()
    }
    var answer = 0L
    val stack = Stack<Int>()

    stack.add(0)

    for (i in 1..n + 1) {
        while (stack.isNotEmpty() && arr[stack.peek()] > arr[i]) {
            val c = stack.peek()
            stack.pop()
            val width = i - stack.peek() - 1
            answer = max(answer, width * arr[c])
        }
        stack.add(i)
    }

    print(answer)
    br.close()
}

File: 백준/Platinum/1725. 히스토그램/히스토그램.cc
#include <stack>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int N;
vector<int> v(100005);
stack<int> st;
int answer;

int main(){
	cin >> N;
	for (int i = 1; i <= N; i++){
		int t; cin >> t;
		v[i] = t;
	}
	st.push(0);
	for (int i = 1; i <= N + 1; i++){
		while (!st.empty() && v[st.top()] > v[i]){
			int c = st.top();
			st.pop();
			answer = max(v[c] * (i - 1 - st.top()), answer);
		}
		st.push(i);
	}
	printf("%d\n", answer);
	return 0;
}